src/analyses/base.ml

```diff
@@ -2030,10 +2031,11 @@ struct
       match exp with
       | None -> nst
       | Some exp ->
-        let t_override = match Cilfacade.fundec_return_type fundec with
-          | TVoid _ -> M.warn ~category:M.Category.Program "Returning a value from a void function"; assert false
-          | ret -> ret
-        in
+        let t_override = Cilfacade.fundec_return_type fundec in
+        if Cil.isVoidType t_override then (
+          M.warn ~category:M.Category.Program "Returning a value from a void function";
+          assert false
+        );
```

Comment:
This is unrelated to the main topic of this PR, but doesn't the `assert false` cause the program to crash, and the warning before to never be printed? Does not seem to make much sense.

Reply:
That might be true indeed if we aren't doing early warnings or such. I'll think about it and perhaps turn it into logging/`failwith`.

Reply:
I just tried triggering this case and couldn't even do it. Apparently CIL already covers it here: https://github.com/goblint/cil/blob/f5ee39bd344dc74e2a10e407d877e0ddf73c9c6f/src/frontc/cabs2cil.ml#L6840-L6843. I think we don't show the CIL warning but CIL removes the returned expression anyway, so Goblint never sees it.

I'll change this to a plain `assert` without a more specific message because this shouldn't even be triggerable by the user.